### PR TITLE
HashOfTag requests cached

### PR DESCRIPTION
**Pull request by EugeneDar**
Now we now cache all requests in HashOfTag.

### CODE REVIEW

**Comment by EugeneDar** [comment]
@mximp please check.

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java [hunk 1]
```diff
@@ -36,6 +38,11 @@
  */
 final class HashOfTag {
 
+    /**
+     * Static map with all hash requests.
+     */
+    private static final Map<String, String> REQUESTED_HASHES = new HashMap<>();
```
Comment (yegor256) [inline eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:44]: @EugeneDar maybe we should make this implementation thread safe? It's important for Maven.

Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:44]: @yegor256 yes, I'll fix it.

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java [hunk 2]
```diff
@@ -56,19 +63,30 @@ final class HashOfTag {
...
             }
         }
-        throw new IllegalArgumentException(
-            String.format("Tag %s doesn't exist in %s", this.tag, link)
-        );
+        if (result.isEmpty()) {
+            throw new IllegalArgumentException(
+                String.format("Tag %s doesn't exist in %s", this.tag, link)
+            );
+        } else {
```
[the remark this answers is not in the document]
Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:86]: @mximp I'll fix it.

[the remark this answers is not in the document]
Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:86]: @mximp I'm sorry, I looked in the wrong place. I'll fix it.

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java (hunk 1 above)
Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:44]: @yegor256 I'll use `ConcurrentHashMap`.

**Comment by EugeneDar** [comment]
@mximp please check again.

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java (hunk 1 above)
Comment (yegor256) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java]: @EugeneDar good idea

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java [hunk 3]
```diff
@@ -56,19 +63,29 @@ final class HashOfTag {
...
-        final InputStream ins = new URL(link).openStream();
-        final Scanner scanner = new Scanner(ins);
-        while (scanner.hasNextLine()) {
-            final String line = scanner.nextLine();
-            final String[] parts = line.split("\t");
-            if (Objects.equals(parts[1], this.tag)) {
-                Logger.info(this, "Git sha of %s is %s", this.tag, parts[0]);
-                return parts[0];
+        String result = "";
+        if (REQUESTED_HASHES.containsKey(this.tag)) {
```
Comment (mximp) [inline eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:67]: @EugeneDar This is still not entirely thread-safe. 
I believe in this case the whole `hash()` method can be marked `synchronised`.

**Review by mximp — CHANGES_REQUESTED** [review]

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java (hunk 3 above)
Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java]: @mximp I fixed it.

**Comment by EugeneDar** [comment]
@mximp please check again.

**Review by mximp — APPROVED** [review]

File: eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java [hunk 4]
```diff
@@ -56,19 +63,31 @@ final class HashOfTag {
      */
     public String hash() throws IOException {
```
Comment (mximp) [inline eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:64]: @EugeneDar maybe it's worth mentioning in JavaDoc for this method that it supports concurrency.

Comment (EugeneDar) [reply eo-maven-plugin/src/main/java/org/eolang/maven/HashOfTag.java:64]: @mximp I will add it.

**Comment by EugeneDar** [comment]
@mximp please check again.

**Comment by EugeneDar** [comment]
@yegor256 please check

**Comment by yegor256** [comment]
@EugeneDar check `ConcurrencyHashMap.computeIfAbsent` method

**Comment by EugeneDar** [comment]
@yegor256 there is no elegant way to throw exception from `computeIfAbsent`, so I suggest to leave initial approach.

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**